juju/url.py

```diff
@@ -20,18 +23,23 @@ def __str__(self):
 
 
 class URL:
+    """Private URL class for this library internals only."""
```

Comment:
```suggestion
    """Private URL class for this library internals only.
    
    Should be instantiated by calling the `URL.parse` constructor.

    If calling `URL` directly without providing the `name` argument,
    the `name` property will be `None`, and must be corrected manually
    before use.
    """
```
`name` argument is a little funky, but your type annotation solution seems good to me. I'd like to add a little bit to the docstring to explain what's going on.